main.cpp

```diff
@@ -0,0 +1,58 @@
+//CodeU: Q1
+//
+//  Created by Julia Semavina on 17.05.2017
+//
+//  problem specification:
+//      Given two strings, write a method to decide if one is a permutation of the other
+//
+//  comments:
+//      register is ignored
+//
+
+#include <iostream>
+#include <map>
+
+
+char lowercase(char a) {
+    if (a <= 'Z' && a >= 'A')
+        return (a - ('Z' - 'z'));
+    return a;
+}
+
+bool is_permutation(std::string &A, std::string &B) {
+    // check if the lengths are equal
+    if (A.length() != B.length())
+        return false;
+    // make a dictionary from string A,
+    // keys - different elements of string A
+    // values - number of each key in string A
+    std::map<char, int> A_elements;
+    char x;
+    for (int i = 0; i < A.length(); ++i) {
+        x = lowercase(A[i]);
+        if (A_elements.find(x) == A_elements.end())
+            A_elements[x] = 1;
+        else
+            A_elements[x] += 1;
+    }
+    // check if string B consists of elements in the dictionary
```

Comment:
Back to this, is usually better to write code than is easier to understand and modify than code that is super optimised. There is tons written about premature optimisation, especially at the cost of readability.
In this case also, the correctness of this second loop depends on the check that both strings are of the same size a few lines above. In practice is very easy for somebody else that comes after you to modify this code to delete that line, and introduce a hard to catch bug.